Core/src/IfcGeometryConverter/CurveConverter.h

```diff
@@ -332,37 +290,124 @@ namespace OpenInfraPlatform {
+					std::vector<carve::geom::vector<3>>& targetVec,
+					std::vector<carve::geom::vector<3>>& segmentStartPoints) const throw(...) 
+				{
+					// the stations at which a point of the tesselation has to be calcuated - to be converted and fill the targetVec
+					std::vector<double> stations = getStationsForTessellationOfIfcAlignmentCurve(alignment_curve);
+
+					carve::geom::vector<3> targetPoint3D;
+					carve::geom::vector<3> targetDirection3D;
+					std::vector<carve::geom::vector<3>> curve_points;
+
+					// attach the curve points
+					for (auto& it_station : stations)
+					{
+						// call the placement converter that handles the geometry and calculates the 3D point along a curve
+						placementConverter->convertBoundedCurveDistAlongToPoint3D(alignment_curve, it_station, true, targetPoint3D, targetDirection3D);
+						curve_points.push_back(targetPoint3D);
+					}
+					GeomUtils::appendPointsToCurve(curve_points, targetVec);
+
+					// add the first point to segments
+					placementConverter->convertBoundedCurveDistAlongToPoint3D(alignment_curve, stations.at(0), true, targetPoint3D, targetDirection3D);
+					segmentStartPoints.push_back(targetPoint3D);
+					// end
+					return;
 				}
 
+
+				/**********************************************************************************************/
+				/*! \brief Calculates curve segments and appends them to the curve.
+				* \param[in] polycurve				A pointer to data from c\ IfcCompositeCurve.
+				* \param[out] targetVec				The tessellated line.
+				* \param[out] segmentStartPoints	The starting points of separate segments.
+				*/
+				void convertIfcCompositeCurve(const EXPRESSReference<typename IfcEntityTypesT::IfcCompositeCurve>& composite_curve,
+					std::vector<carve::geom::vector<3>>& targetVec,
+					std::vector<carve::geom::vector<3>>& segmentStartPoints) const throw(...)
+				{
+
+					std::vector<std::shared_ptr<typename IfcEntityTypesT::IfcCompositeCurveSegment> > segments;
+					segments.resize(composite_curve->Segments.size());
+
+					std::transform(
+						composite_curve->Segments.begin(),
+						composite_curve->Segments.end(),
+						segments.begin(),
+						[](auto &it) { return it.lock(); });
+
+					std::vector<std::shared_ptr<typename IfcEntityTypesT::IfcCompositeCurveSegment> >::iterator it_segments =
+						segments.begin();
+
+					for (; it_segments != segments.end(); ++it_segments) {
+						std::shared_ptr<typename IfcEntityTypesT::IfcCompositeCurveSegment> segment = (*it_segments);
+						std::shared_ptr<typename IfcEntityTypesT::IfcCurve> segment_curve = segment->ParentCurve.lock();
+
+						std::vector<carve::geom::vector<3>> segment_vec;
+						convertIfcCurve(segment_curve, segment_vec, segmentStartPoints);
+						if (!segment_vec.empty()) {
+							GeomUtils::appendPointsToCurve(segment_vec, targetVec);
+						}
+					}
+					// end
+					return;

```

Comment:
This can be greatly simplified with the use of `EXPRESSReference`.

Reply:
Resolved:  4878063